secant/Util/RecoveryPhraseStorage.swift

```diff
@@ -0,0 +1,228 @@
+        }
+    }
+    
+    /// Restore data for key
+    func data(
+        forKey: String,
+        account: String = ""
+    ) -> Data? {
+        let query = RecoveryPhraseStorage.restoreQuery(forAccount: account, andKey: forKey)
+
+        var result: AnyObject?
+        _ = SecItemCopyMatching(query as CFDictionary, &result)
+        
+        return result as? Data
+    }
+    
+    /// Use carefully:  Deletes data for key
+    func deleteData(
+        forKey: String,
+        account: String = ""
+    ) -> Bool {
+        let query = RecoveryPhraseStorage.baseQuery(forAccount: account, andKey: forKey)
+
+        let status = SecItemDelete(query as CFDictionary)
+
+        return status == noErr
+    }
+    
+    /// Store data for key
+    func setData(
+        _ data: Data,
+        forKey: String,
+        account: String = ""
+    ) throws {
+        var query = RecoveryPhraseStorage.baseQuery(forAccount: account, andKey: forKey)
+        query[kSecValueData as String] = data as AnyObject
+
+        let status = SecItemAdd(query as CFDictionary, nil)
+        
+        guard status != errSecDuplicateItem else {
+            throw KeychainError.duplicate
+        }
+        
+        guard status == errSecSuccess else {
+            throw KeychainError.unknown(status)
+        }
+    }
+}
+
+// MARK: - KeyStoring
+
+extension RecoveryPhraseStorage: KeyStoring {
+    func importBirthday(_ height: BlockHeight) throws {
+        do {
+            try setBirthday(height)
+        } catch RecoveryPhraseStorage.KeychainError.duplicate {
+            throw KeyStoringError.alreadyImported
+        } catch {
+            throw KeyStoringError.storageError(error)
+        }
+    }
+    
+    func exportBirthday() throws -> BlockHeight {
+        do {
+            return try birthday()
+        } catch {
+            throw KeyStoringError.uninitializedWallet
+        }
+    }
+    
+    func importPhrase(bip39 phrase: String) throws {
+        do {
+            try setRecoveryPhrase(phrase)
+        } catch RecoveryPhraseStorage.KeychainError.duplicate {
+            throw KeyStoringError.alreadyImported
+        } catch {
+            throw KeyStoringError.storageError(error)
+        }
+    }
+    
+    func exportPhrase() throws -> String {
+        do {
+            return try recoveryPhrase()
+        } catch {
+            throw KeyStoringError.uninitializedWallet
+        }
+    }
+    
+    func areKeysPresent() throws -> Bool {
+        do {
+            _ = try exportPhrase()
+            _ = try exportBirthday()
+        } catch KeyStoringError.uninitializedWallet {
+            return false
+        } catch {
+            // TODO: - report & log error.localizedDescription
+            return false
+        }
+        
+        return true
+    }
+    
+    func nukePhrase() {
+        deleteRecoveryPhrase()
+    }
+    
+    func nukeBirthday() {
+        deleteBirthday()
+    }
+    
+    func nukeWallet() {
```

Comment:
I'm concerned about atomicity here.  I think we might have a problem if a crash happens between `nukePhrase` and `nukeBirthday`.  Assuming that is true, some possible mitigations:
 - Re-order the delete so that we nuke the birthday first, then the phrase.  A nuked birthday is recoverable but a nuked phrase is not.
 - We store phrase and birthday as a single keychain item.  This guarantees the two items are always written atomically.  This is what the Android wallet is doing currently, although that doesn't necessarily mean it is the ideal design.

Reply:
I find this suggestion really useful. Can you point us to the implementation of that so we can match it?

Reply:
This is the Android implementation

Key-value preference store:
https://github.com/zcash/secant-android-wallet/blob/main/preference-impl-android-lib/src/main/java/co/electriccoin/zcash/preference/AndroidPreferenceProvider.kt

Serialization of wallet: (note that I will might want to change my format to be storing a `BirthdayHeight` instead of a `Birthday` since the `Birthday` has extra fields that I'm not sure we need to persist)
https://github.com/zcash/secant-android-wallet/blob/main/sdk-ext-lib/src/main/java/cash/z/ecc/sdk/model/PersistableWallet.kt

UI entry point to write the secret:
https://github.com/zcash/secant-android-wallet/blob/bb43d2f7a612b01ef0256bff3a45c932c6d44e60/ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/home/viewmodel/WalletViewModel.kt#L152